raftstore panic when staring TiKV

## Bug Report

**What did you do?**

Just start TiKV, panic. but after panic several times, the TiKV can start successfully. 

**What did you expect to see?**

No panic

**What did you see instead?**

```
[2019/02/16 20:46:32.466 +08:00] [ERROR] [mod.rs:475] ["thread 'raftstore-6-0' panicked 'called `Option::unwrap()` on a `None` value' at \"src/libcore/option.rs:355\"\nstack backtrace:\n   0:     0x55b0eda0898d - backtrace::backtrace::libunwind::trace::h42bd35238981daad\n                        at /home/qupeng/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.2.3/src/backtrace/libunwind.rs:54\n                         - backtrace::backtrace::trace::h1f00fe0b81b1621d\n                        at /home/qupeng/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.2.3/src/backtrace/mod.rs:70\n   1:     0x55b0ed38c75b - tikv::util::set_panic_hook::{{closure}}::h9c9bb177cf30c18e\n                        at /home/qupeng/github.com/pingcap/tikv/target/release/build/backtrace-ce050514236e0111/out/capture.rs:79\n   2:     0x55b0edaed9fd - std::panicking::rust_panic_with_hook::he9362bdf637fc5d8\n                        at src/libstd/panicking.rs:495\n   3:     0x55b0edaed7be - std::panicking::continue_panic_fmt::h8c8ecd2074b327b3\n                        at src/libstd/panicking.rs:398\n   4:     0x55b0edb01ef5 - rust_begin_unwind\n                        at src/libstd/panicking.rs:325\n   5:     0x55b0edb0685b - core::panicking::panic_fmt::hc84bf3901c1095d4\n                        at src/libcore/panicking.rs:95\n   6:     0x55b0edb0831a - core::panicking::panic::h2cd043862cc6dc4a\n                        at src/libcore/panicking.rs:59\n   7:     0x55b0ed26bd46 - <core::option::Option<T>>::unwrap::h6f447e9a70a6d3cf\n                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libcore/macros.rs:20\n                         - <tikv::raftstore::store::fsm::store::StoreFsmDelegate<'a, T, C>>::store_heartbeat_pd::h14cd37177d1aa0fd\n                        at /home/qupeng/github.com/pingcap/tikv/src/raftstore/store/fsm/store.rs:1534\n   8:     0x55b0ed2631ef - <tikv::raftstore::store::fsm::store::StoreFsmDelegate<'a, T, C>>::handle_msgs::h1cb2dce6bf1dd9be\n                        at /home/qupeng/github.com/pingcap/tikv/src/raftstore/store/fsm/store.rs:452\n   9:     0x55b0ed1e944a - <tikv::raftstore::store::fsm::store::RaftPoller<T, C> as tikv::raftstore::store::fsm::batch::PollHandler<tikv::raftstore::store::fsm::peer::PeerFsm, tikv::raftstore::store::fsm::store::StoreFsm>>::handle_control::h953defd5ec158028\n                        at /home/qupeng/github.com/pingcap/tikv/src/raftstore/store/fsm/store.rs:615\n                         - <tikv::raftstore::store::fsm::batch::Poller<N, C, Handler>>::poll::hfc9fcd194235867e\n                        at /home/qupeng/github.com/pingcap/tikv/src/raftstore/store/fsm/batch.rs:334\n  10:     0x55b0ed007153 - <tikv::raftstore::store::fsm::batch::BatchSystem<N, C>>::spawn::{{closure}}::h4216ac29e8293fd7\n                        at /home/qupeng/github.com/pingcap/tikv/src/raftstore/store/fsm/batch.rs:412\n                         - std::sys_common::backtrace::__rust_begin_short_backtrace::h88070bb57e05ea4a\n                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libstd/sys_common/backtrace.rs:136\n  11:     0x55b0ed14ff20 - std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}::hfd8153758c5de686\n                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libstd/thread/mod.rs:477\n                         - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::hf66778517c33afb3\n                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libstd/panic.rs:319\n                         - std::panicking::try::do_call::h9fc881c74d130201\n                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libstd/panicking.rs:310\n                         - std::panicking::try::h32753ac12e5d4ef1\n                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56//src/libpanic_unwind/lib.rs:102\n                         - std::panic::catch_unwind::h39e4a909b32d19ce\n                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libstd/panic.rs:398\n                         - std::thread::Builder::spawn_unchecked::{{closure}}::h7c9247b83564e7d6\n                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libstd/thread/mod.rs:476\n                         - <F as alloc::boxed::FnBox<A>>::call_box::h74fc0ea5bd0b31f6\n                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/liballoc/boxed.rs:673\n  12:     0x55b0edb03a9d - <alloc::boxed::Box<(dyn alloc::boxed::FnBox<A, Output=R> + 'a)> as core::ops::function::FnOnce<A>>::call_once::hb8f4ee9bdf46e5d9\n                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/liballoc/boxed.rs:683\n                         - std::sys_common::thread::start_thread::h938451c06a28c9dd\n                        at src/libstd/sys_common/thread.rs:24\n                         - std::sys::unix::thread::Thread::new::thread_start::h809dab03f76374f4\n                        at src/libstd/sys/unix/thread.rs:90\n  13:     0x7f202c7b6dd4 - start_thread\n  14:     0x7f202bbb6eac - __clone\n  15:                0x0 - <unknown>"]
[2019/02/16 20:46:32.466 +08:00] [INFO] [unknown:920] ["[region 3201491] 3201493 [term: 8] received a MsgHeartbeat message with higher term from 3201492 [term: 9]"]
```

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -85,6 +85,7 @@ extern crate serde_json;
     slog_warn,
     slog_info,
     slog_debug,
+    slog_crit,
     slog_log,
     slog_record,
     slog_b,

--- a/src/server/server.rs
+++ b/src/server/server.rs
@@ -31,6 +31,7 @@ use crate::raftstore::store::{Engines, SnapManager};
 use crate::storage::{Engine, Storage};
 use crate::util::security::SecurityManager;
 use crate::util::worker::Worker;
+use crate::util::Either;
 
 use super::load_statistics::*;
 use super::raft_client::RaftClient;
@@ -52,8 +53,10 @@ pub const STATS_THREAD_PREFIX: &str = "transport-stats";
 /// and a snapshot worker.
 pub struct Server<T: RaftStoreRouter + 'static, S: StoreAddrResolver + 'static> {
     env: Arc<Environment>,
-    // Grpc server.
-    grpc_server: GrpcServer,
+    /// A GrpcServer builder or a GrpcServer.
+    ///
+    /// If the listening port is configured, the server will be started lazily.
+    builder_or_server: Option<Either<ServerBuilder, GrpcServer>>,
     local_addr: SocketAddr,
     // Transport.
     trans: ServerTransport<T, S>,
@@ -107,7 +110,7 @@ impl<T: RaftStoreRouter, S: StoreAddrResolver + 'static> Server<T, S> {
             Arc::clone(&thread_load),
         );
 
-        let addr = SocketAddr::from_str(&cfg.addr)?;
+        let mut addr = SocketAddr::from_str(&cfg.addr)?;
         info!("listening on {}", addr);
         let ip = format!("{}", addr.ip());
         let channel_args = ChannelBuilder::new(Arc::clone(&env))
@@ -117,7 +120,7 @@ impl<T: RaftStoreRouter, S: StoreAddrResolver + 'static> Server<T, S> {
             .max_send_message_len(-1)
             .http2_max_ping_strikes(i32::MAX) // For pings without data from clients.
             .build_args();
-        let grpc_server = {
+        let builder_or_server = {
             let mut sb = ServerBuilder::new(Arc::clone(&env))
                 .channel_args(channel_args)
                 .register_service(create_tikv(kv_service));
@@ -129,12 +132,14 @@ impl<T: RaftStoreRouter, S: StoreAddrResolver + 'static> Server<T, S> {
             if let Some(service) = import_service {
                 sb = sb.register_service(create_import_sst(service));
             }
-            sb.build()?
-        };
-
-        let addr = {
-            let (ref host, port) = grpc_server.bind_addrs()[0];
-            SocketAddr::new(IpAddr::from_str(host)?, port as u16)
+            if addr.port() == 0 {
+                let server = sb.build()?;
+                let (ref host, port) = server.bind_addrs()[0];
+                addr = SocketAddr::new(IpAddr::from_str(host)?, port as u16);
+                Either::Right(server)
+            } else {
+                Either::Left(sb)
+            }
         };
 
         let raft_client = Arc::new(RwLock::new(RaftClient::new(
@@ -154,7 +159,7 @@ impl<T: RaftStoreRouter, S: StoreAddrResolver + 'static> Server<T, S> {
 
         let svr = Server {
             env: Arc::clone(&env),
-            grpc_server,
+            builder_or_server: Some(builder_or_server),
             local_addr: addr,
             trans,
             raft_router,
@@ -181,7 +186,13 @@ impl<T: RaftStoreRouter, S: StoreAddrResolver + 'static> Server<T, S> {
             Arc::clone(&cfg),
         );
         box_try!(self.snap_worker.start(snap_runner));
-        self.grpc_server.start();
+        let builder_or_server = self.builder_or_server.take().unwrap();
+        let mut grpc_server = match builder_or_server {
+            Either::Left(builder) => builder.build()?,
+            Either::Right(server) => server,
+        };
+        grpc_server.start();
+        self.builder_or_server = Some(Either::Right(grpc_server));
 
         let mut load_stats = {
             let tl = Arc::clone(&self.thread_load);
@@ -203,7 +214,9 @@ impl<T: RaftStoreRouter, S: StoreAddrResolver + 'static> Server<T, S> {
     /// Stops the TiKV server.
     pub fn stop(&mut self) -> Result<()> {
         self.snap_worker.stop();
-        self.grpc_server.shutdown();
+        if let Some(Either::Right(mut server)) = self.builder_or_server.take() {
+            server.shutdown();
+        }
         Ok(())
     }
 

--- a/src/server/server.rs
+++ b/src/server/server.rs
@@ -132,6 +132,8 @@ impl<T: RaftStoreRouter, S: StoreAddrResolver + 'static> Server<T, S> {
             if let Some(service) = import_service {
                 sb = sb.register_service(create_import_sst(service));
             }
+            // When port is 0, it has to be binded now to get a valid address, which
+            // is then reported to PD before the server is up. 0 is usually used in tests.
             if addr.port() == 0 {
                 let server = sb.build()?;
                 let (ref host, port) = server.bind_addrs()[0];

--- a/src/util/mod.rs
+++ b/src/util/mod.rs
@@ -472,12 +472,10 @@ pub fn set_panic_hook(panic_abort: bool, data_dir: &str) {
                 .location()
                 .map(|l| format!("{}:{}", l.file(), l.line()));
             let bt = backtrace::Backtrace::new();
-            error!(
-                "thread '{}' panicked '{}' at {:?}\n{:?}",
-                name,
-                msg,
-                loc.unwrap_or_else(|| "<unknown>".to_owned()),
-                bt
+            crit!("{}", msg;
+                "thread_name" => name,
+                "location" => loc.unwrap_or_else(|| "<unknown>".to_owned()),
+                "backtrace" => format_args!("{:?}", bt),
             );
         } else {
             orig_hook(info);

